### PR TITLE
[flashpoint-163] Flashpoint Feed Integration

### PR DESCRIPTION
**Pull request by crestdatasystems**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Flashpoint is the globally trusted leader in risk intelligence for the fastest, most comprehensive coverage of threatening activity on the internet. From bolstering cyber and physical security, to detecting fraud and insider threats, Flashpoint partners with customers across the private and public sectors to help them rapidly identify threats and mitigate their most critical security risks.

## Minimum version of Cortex XSOAR
- [ ] 5.5.0
- [x] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/FlashpointFeed/Classifiers/classifier-mapper-incoming-Flashpoint_Feed.json [hunk 1]
```diff
@@ -0,0 +1,690 @@
+{
```
Comment (ChanochShayner) [inline Packs/FlashpointFeed/Classifiers/classifier-mapper-incoming-Flashpoint_Feed.json:1]: Currently, in Feed Integrations, we map the indicator fields within the integration code (so we can handle the information that comes back most conveniently, and also save Classifier configuration)
Is there a specific reason you created the Classifier?

File: Packs/FlashpointFeed/IndicatorFields/incidentfield-Flashpoint-Attribute-ID.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    "required": false,
+    "neverSetAsRequired": false,
+    "isReadOnly": false,
+    "useAsKpi": false,
+    "locked": false,
+    "system": false,
+    "content": true,
+    "group": 2,
+    "hidden": false,
+    "associatedToAll": true,
```
Comment (ChanochShayner) [inline Packs/FlashpointFeed/IndicatorFields/incidentfield-Flashpoint-Attribute-ID.json:21]:
```suggestion
    "associatedToAll": false,
    "associatedTypes": [
        ""Flashpoint Indicator""
    ],
```
The same change in all other Indicator Fields

File: Packs/FlashpointFeed/Layouts/layoutscontainer-Flashpoint_Indicator.json [hunk 3]
```diff
@@ -0,0 +1,386 @@
+{
```
Comment (ChanochShayner) [inline Packs/FlashpointFeed/Layouts/layoutscontainer-Flashpoint_Indicator.json:1]: Please add the Relationships section to the layout

File: Packs/FlashpointFeed/README.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+The following types of indicators are retrieved:
+
+- IP
+- Domain
+- URL
+- Email
+- SHA1
+- SHA256
+- MD5
```
Comment (ChanochShayner) [inline Packs/FlashpointFeed/README.md:17]:
```suggestion
- File
```
We currently only produce File indicator type and not SHA1 and SHA256 and MD5
And this is also what the `auto_detect_indicator_type` function you use does.

File: Packs/FlashpointFeed/pack_metadata.json [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    "support": "partner",
+    "currentVersion": "1.0.0",
+    "author": "Flashpoint",
+    "url": "https://www.flashpoint-intel.com/",
+    "email": "demisto_support@flashpoint-intel.com",
+    "categories": [
+        "Data Enrichment & Threat Intelligence"
+    ],
+    "tags": [
+        "Threat Intelligence"
```
Comment (ChanochShayner) [inline Packs/FlashpointFeed/pack_metadata.json:13]:
```suggestion
        "Threat Intelligence",
        "Relationship",
        "TIM"
```

File: Packs/FlashpointFeed/pack_metadata.json [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    ],
+    "useCases": ["Threat Intelligence Management"],
+    "keywords": ["Flashpoint", "Indicators", "IoC"],
```
Comment (ChanochShayner) [inline Packs/FlashpointFeed/pack_metadata.json:16]:
```suggestion
    "keywords": ["Flashpoint", "Indicators", "IoC", "Feed"],
```

**Review by ChanochShayner — COMMENTED** [review]
looks great! Thanks!
Please fill out the Contribution Registration form and the Contributor License Agreement.
And pay attention to my comments.

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+        :return: Type of the indicator.
+        """
+        ind_type = DEFAULT_INDICATOR_TYPE
+        if not default_map:
+            ind_type = auto_detect_indicator_type(indicator_value)
+
+            if not ind_type:
+                ind_type = DEFAULT_INDICATOR_TYPE
+
+        return ind_type
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py:82]: Hey @ChanochShayner 
We along with the customer agreed to settle with `Flashpoint Indicator` as they wanted us to include other indicators in Cortex XSOAR.
Also, we confirmed with the XSOAR team what to do in such cases and they also proposed to create a custom indicator.

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+
+# Disable insecure warnings
+requests.packages.urllib3.disable_warnings()  # pylint: disable=no-member
+
+''' CONSTANTS '''
+
+DATE_FORMAT = '%Y-%m-%dT%H:%M:%SZ'  # ISO8601 format with UTC, default in XSOAR
+LIMIT = 10
+TIMEOUT = 60
+MAX_FETCH = 1000
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py:16]: We were getting a timeout error randomly from Flashpoint when MAX_FETCH is >2000 which we conveyed to the customer and agreed to set 1000 as the MAX_FETCH value.

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+    """Tests the indicators from the feed.
+    Args:
+        client: Client object.
+    Returns:
+        Outputs.
+    """
+    client.http_request(url_suffix='/indicators/attribute', params={"limit": 1})
+    is_fetch = params.get('feed')
+    if is_fetch:
+        fetch_indicators_command(client, params, {}, True, False)
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py:307]: As per the demisto's standard, if any integration is having to fetch incident/indicator functionality then it should be tested when the "Test" button is hit.
In this case, the default value is 'True' for fetch indicator but customer can still enable or disable the fetch indicator functionality. 
The purpose of passing `is_fetch` value as True in  `fetch_incdicators_command` is to validate the fetch indicator-related setting such as "First fetch time interval".

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+            entity_b = tag.get('name')
+            if entity_b:
+                obj = EntityRelationship(
+                    name=EntityRelationship.Relationships.INDICATOR_OF,
+                    entity_a=entity_a,
+                    entity_a_type=entity_a_type,
+                    entity_b=entity_b,
+                    entity_b_type='Flashpoint Indicator',
+                    brand="FlashpointFeed"
+                )
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py:108]: sure @ChanochShayner  will make this change.

File: Packs/FlashpointFeed/Classifiers/classifier-mapper-incoming-Flashpoint_Feed.json (hunk 1 above)
Comment (crestdatasystems) [reply Packs/FlashpointFeed/Classifiers/classifier-mapper-incoming-Flashpoint_Feed.json:1]: We have classified the indicator through code (generic and custom indicator) and wanted to have fewer dependencies on the code so we created a mapper to map the indicator fields.

File: Packs/FlashpointFeed/Layouts/layoutscontainer-Flashpoint_Indicator.json (hunk 3 above)
Comment (crestdatasystems) [reply Packs/FlashpointFeed/Layouts/layoutscontainer-Flashpoint_Indicator.json:1]: sure @ChanochShayner, it seems it is recently released. Will do it anyway.

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.yml [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+  defaultvalue: 'true'
+  hidden: true
+  name: feedIncremental
+  required: false
+  type: 8
+- additionalinfo: Create relationships between indicators as part of Enrichment.
+  defaultvalue: 'false'
+  display: Relationship
+  hidden: false
+  name: createRelationship
```
Comment (guykeller) [inline Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.yml:116]: let's change to `create_relationship`, and the same goes for all other names (`defaultMap` for example)

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.yml [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+    provides new indicators for the last day or a feed which is immutable and provides
+    indicators from a search date onwards.
+  defaultvalue: 'true'
+  hidden: true
+  name: feedIncremental
+  required: false
+  type: 8
+- additionalinfo: Create relationships between indicators as part of Enrichment.
+  defaultvalue: 'false'
+  display: Relationship
```
Comment (guykeller) [inline Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.yml:114]: This name should possibly be improved
@ChanochShayner do you think it's clear enough?

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (guykeller) [reply Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py:307]: I tend to agree with @ChanochShayner here
Currently, we aren't actually testing anything
We will probably be wise to actually test the logic at some level

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.yml (hunk 11 above)
Comment (ChanochShayner) [reply Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.yml:116]: @guykeller The names of these parameters are Mandatory, they are supposed to be as they are here and cannot be changed.

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.yml (hunk 12 above)
Comment (ChanochShayner) [reply Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.yml:114]: Same as above.

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py:307]: Noted

**Comment by crestdatasystems** [comment]
@ChanochShayner Please assist us with the `Packs/FlashpointFeed/Author_image.png` error on CICD.

**Comment by crestdatasystems** [comment]
@ChanochShayner Please note: below files came after taking the latest changes from remote/master branch

Packs/CommonScripts/.secrets-ignore
Packs/CommonScripts/ReleaseNotes/1_4_33.md 
Packs/CommonScripts/Scripts/ParseEmailFiles/ParseEmailFiles.py 
Packs/CommonScripts/Scripts/ParseEmailFiles/ParseEmailFiles.yml
Packs/CommonScripts/Scripts/ParseEmailFiles/ParseEmailFiles_test.py

File: Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+                if 'misp-galaxy:mitre-enterprise-attack' in entity_b:
+                    name = entity_b.split("=")
+                    entity_b = name[1].replace('\"', "")
+
+                obj = EntityRelationship(
+                    name=EntityRelationship.Relationships.INDICATOR_OF,
+                    entity_a=entity_a,
+                    entity_a_type=entity_a_type,
+                    entity_b=entity_b,
+                    entity_b_type='Attack Pattern'
```
Comment (ChanochShayner) [inline Packs/FlashpointFeed/Integrations/FlashpointFeed/FlashpointFeed.py:111]: The Indicator type in entity_b_type will not always be of the `Attack Pattern` type, there may be other Threat Intel Objs such as Malware, Tool, Intrusion Set etc.

I think it's supposed to look like this:
"misp-galaxy:mitre-enterprise-attack-attack-pattern=........"
"misp-galaxy:mitre-enterprise-attack-Malware=......"
"misp-galaxy:mitre-enterprise-attack-Tool=........"

Please check it out.

File: Packs/FlashpointFeed/Classifiers/classifier-mapper-incoming-Flashpoint_Feed.json (hunk 1 above)
Comment (ChanochShayner) [reply Packs/FlashpointFeed/Classifiers/classifier-mapper-incoming-Flashpoint_Feed.json:1]: If there is no special reason, then please add the mapping of the fields within the code (as in all the feeds we have).

**Comment by crestdatasystems** [comment]
hi @ShahafBenYakir,
Let us know if we have something left in terms of PR updates.

**Comment by guykeller** [comment]
Due to a sequence of holidays in recent weeks, your review was delayed. We apologize for the inconvenience and will try to do our best to complete your review as quickly and efficiently as possible, please bare with us.
Thanks for your understanding, and for your contribution.

**Review by ChanochShayner — APPROVED** [review]

**Pull request merged**